objects/src/accounts/storage.rs

```diff
@@ -1,47 +1,83 @@
+/// Account storage is composed of two components. The first component is a simple sparse Merkle
+/// tree of depth 8 which is index addressable. This provides the user with 256 Word slots. Users
+/// that require additional storage can use the second component which is a `MerkleStore`. This
+/// will allow the user to store any Merkle structures they need.  This is achieved by storing the
+/// root of the Merkle structure as a leaf in the simple sparse merkle tree. When `AccountStorage`
+/// is serialized it will check to see if any of the leafs in the simple sparse Merkle tree are
+/// Merkle roots of other Merkle structures.  If any Merkle roots are found then the Merkle
+/// structures will be persisted in the `AccountStorage` `MerkleStore`.
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct AccountStorage {
-    // TODO: add compact SMT as backing storage
+    slots: SimpleSmt,
+    store: MerkleStore,
 }
 
 impl AccountStorage {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// Depth of the storage tree.
+    pub const STORAGE_TREE_DEPTH: u8 = 8;
+
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Returns a new instance of account storage initialized with the provided items.
-    pub fn new(_items: &[StorageItem]) -> Self {
-        Self {}
+    pub fn new(
+        items: Vec<StorageItem>,
+        store: MerkleStore,
+    ) -> Result<AccountStorage, AccountError> {
+        let slots = SimpleSmt::new(Self::STORAGE_TREE_DEPTH)
+            .expect("depth is valid")
+            .with_leaves(items.into_iter().map(|x| (x.0 as u64, x.1)))
+            .expect("`StorageItem` index is u8 - index within range");
+
+        Ok(Self { slots, store })
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
     /// Returns a commitment to this storage.
-    pub fn root(&self) -> Digest {
-        Digest::default()
+    pub fn root(&self) -> Word {
+        self.slots.root()
     }
 
-    /// Returns an item from the storage at the specified key. Both items and keys are Words which
-    /// consist of 4 field elements.
+    /// Returns an item from the storage at the specified index.
     ///
     /// If the item is not present in the storage, [ZERO; 4] is returned.
-    pub fn get_item(&self, key: Word) -> Word {
-        // the index of the item is the hash of its key
-        let _index: Word = Hasher::merge(&[key.into(), [ZERO; 4].into()]).into();
-        todo!("retrieve the item from the tree");
+    pub fn get_item(&self, index: u8) -> Result<Word, AccountError> {
+        let item_index =
+            NodeIndex::new(Self::STORAGE_TREE_DEPTH, index as u64).expect("u8 is valid");
+        self.slots.get_node(item_index).map_err(AccountError::GetStorageItemError)
+    }
```

Comment:
Under what circumstances would this return an error? Seems like the only reason for `SimpleSmt::get_node()` to return an error would be if the index has issues, but I think we can guarantee that this shouldn't happen.

Maybe this method should be infallible?

Reply:
so the branch we are interested in from `get_node` is:

```rust
        ...} else if index.depth() == self.depth() {
            self.get_leaf_node(index.value())
                .or_else(|| self.empty_hashes.get(index.depth() as usize).copied().map(Word::from))
                .ok_or(MerkleError::NodeNotInSet(index.value()))
        } else {...
```

This doesn't appear to be infallible and hence I propagated the errors.  I'm not sure under what circumstances (if any) this would actually fail in our case (or any case for that matter)?

Reply:
I think this would fail only if there is a bug in the constructor of `SimpleSmt`. So, maybe this should be converted into an `expect` or something like that.

Reply:
Yeah I think that makes sense.

Reply:
updated